Add native unit types for weather entities

rianadon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
I'm attempting to port @emontnemery's [addition of `native_units` to SensorEntity](https://github.com/home-assistant/core/pull/48261) to the `WeatherEntity` class.

This will allow integrations publishing weather data results to convert their data to local units, without needing to write unit conversion routines in each integration. This can also be used as a base to implement unit conversion for weather templates, like in #48641.

It is very unlikely this change should break custom components. However, I did find one component (`met_office`) that already defined a `visibility_unit` method but did not return floats in its `visibility` method. If any other components pull similar shenanigans, they will break.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rianadon committed:
Add native unit types for weather entities

emontnemery commented:
I think this looks good!

To prevent the `@final` decorator from breaking CI: don't add it until all existing integrations are updated.

emontnemery commented:
Tests are needed too, I just noticed it's missing from sensor tests.
Maybe do something like this: https://github.com/home-assistant/core/pull/59546

rianadon committed:
Update weatherentity and change precision in climacell test

rianadon committed:
Move weather test to demo tests

rianadon committed:
Add weather test for temperature conversion

rianadon committed:
Add more unit conversion tests

rianadon commented:
I've added a bit more. There are tests, and also as a result of cleaning up unit types the precision of temperature values is no longer set by the integration's units but rather the user's units. So I updated the climacell test to add the precision necessary for Celsius values.

rianadon committed:
Remove extra native_ methods

rianadon committed:
Remove extra properties and save precision change for another PR

rianadon commented:
I'm changing the names of the new methods to no longer have the `native_` prefixes in front of them to be more consistent with what we have already. Adding a separate `native_temperature` also made it difficult to decide whether to use the integration's overridden `temperature` method or to use the default `native_temperature` implementation.

Plus, everything gets unit-converted anyways in `state_attributes`.

rianadon committed:
Remove visibility_unit from metoffice component

The vibility values given by metoffice are formatted into strings,
which means they can't automatically be converted.

emontnemery commented on homeassistant/components/weather/__init__.py:
```diff
@@ -126,14 +129,19 @@ def temperature(self) -> float | None:
     @property
     def temperature_unit(self) -> str:
-        """Return the unit of measurement."""
+        """Return the unit of measurement for temperature."""
```
Let's make it clear in the docstring this is the weather entity's unit, before conversion, same for the other units.

emontnemery commented on homeassistant/components/weather/__init__.py:
```diff
@@ -182,7 +200,10 @@ def state_attributes(self):
         data = {}
```
Update the docstring to explain the unit conversion.

emontnemery commented on homeassistant/components/weather/__init__.py:
```diff
@@ -144,6 +152,11 @@ def wind_speed(self) -> float | None:
+    @property
+    def wind_speed_unit(self) -> str | None:
+        """Return the unit of measurement for wind speed."""
+        return self._attr_wind_speed_unit
```
This seems to be missed in the tests?

rianadon replied:
My mistake! I forgot to mention I'm waiting for #58437 to be merged so that we know which units to convert to.

rianadon committed:
Improve docstrings and convert pressures in forecast

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into weather-unit-conversion

rianadon committed:
Add precipitation and wind speed units

rianadon committed:
Clean up tests

rianadon commented:
The code should be ready to go now!

It looks like the test is failing due to Hue bridge stuff which is strange.

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into weather-unit-conversion

rianadon committed:
Round converted weather values

rianadon committed:
Round weather values to 2 decimal places

rianadon committed:
Move number of rounding decimal places to constant

emontnemery commented on homeassistant/components/weather/__init__.py:
```diff
@@ -112,38 +114,52 @@ class WeatherEntity(Entity):
     @property
     def temperature(self) -> float | None:
-        """Return the platform temperature."""
+        """Return the platform temperature in native units (i.e. not converted)."""
```
For clarity, let's have an elaborate docstring similar to this one also on the other properties.

emontnemery commented on homeassistant/components/weather/__init__.py:
```diff
@@ -61,6 +61,8 @@
 
 SCAN_INTERVAL = timedelta(seconds=30)
 
+ROUNDING_PLACES = 2
```
Let's call it something `_PRECISION`

emontnemery commented on tests/testing_config/custom_components/test/weather.py:
```diff
@@ -0,0 +1,126 @@
+"""
+Provide a mock sensor platform.
```
```suggestion
Provide a mock weather platform.
```

rianadon committed:
Docstring and styles

emontnemery reviewed (approved):
Thanks, @rianadon 👍

frenck commented on homeassistant/components/weather/__init__.py:
```diff
@@ -192,15 +221,28 @@ def state_attributes(self):
         if (wind_speed := self.wind_speed) is not None:
+            if (unit := self.wind_speed_unit) is not None:
+                wind_speed = round(
+                    self.hass.config.units.wind_speed(wind_speed, unit),
```
Why is hass config doing the unit conversion?

`wind_seed` should be imported directly from `util.unit_system` instead.

The same applies to all other unit conversions happening here.

frenck commented:
This PR is causing massive breaking changes. 

While I think the idea is good, I think the way it is set up is not fitting nor correct.

For that reason:

- Opened up #63842 to revert #61472
- Opened up #63843 to revert #61470

I think it is cool to support unit conversion on an entity or device level in the future, but we cannot rely on core configuration for e.g., unit of mass or pressure. Which unit to use for those cases heavily relies on the use case.
